docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+ - Min / Max refill amount. These are just safeguards to keep sane behavior in
+   potential corner-cases.
+
+### Prototype
+
+We created a simulation and visualization framework in order to prototype this
+algorithm.  The input is a workload, which is a set of requested RU/s graphs
+(one per node). The output is a set of granted RU/s graphs. The simulation
+implements this algorithm and compares against an ideal distributed token
+bucket.
+
+The simulation can be accessed at:
+[https://raduberinde.github.io/distbucket-v21-06-14/?workload=steps](https://raduberinde.github.io/distbucket-v21-06-14/?workload=steps).
+
+The page shows the granted workload over time for both the prototyped algorithm
+and an ideal (instantly synchronized) token bucket. We also show the graphs of
+total consumption between the two, to see if there is any systematic (long-term)
+divergence.
+
+The workload can be changed from the dropdown or the yaml can be manually
+edited. Each workload function is a sum of terms; the supported terms can be
+inferred from the
+[code](https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/data.go#L77).
+
+The code is available here:
+[https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/dist_token_bucket_3.go](https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/dist_token_bucket_3.go).
+
+## System table
+
+The system table uses a "ledger" approach; each change to the state of a global
+tenant bucket appends a row to the table. Each row has a unique operation UUID
+and a sequence number which is unique within that tenant. The operation UUID
+allows detection of duplicate requests, to ensure idempotency. The sequence
+numbers allow locating the latest entry for each tenant.
+
+Schema for the system table:
+```sql
+CREATE TABLE tenant_usage (
+  tenant_id INT NOT NULL,
+  seq INT NOT NULL,
+
+  op_id UUID NOT NULL,  -- TODO: maybe INT?
+  op_time TIMESTAMP NOT NULL,
+
+  ru_burst_limit FLOAT NOT NULL,
+  ru_rate FLOAT NOT NULL,
+
+  ru_current FLOAT NOT NULL,
+  current_share_sum FLOAT NOT NULL,
+
+  -- Cumulative usage statistics.
+  total_ru_usage FLOAT NOT NULL,
+  total_read_requests INT NOT NULL,
+  total_read_bytes INT NOT NULL,
+  total_write_requests INT NOT NULL,
+  total_write_bytes INT NOT NULL,
+  total_sql_pod_cpu_seconds FLOAT NOT NULL, -- TODO: Maybe milliseconds and INT8?
+
+  PRIMARY KEY (tenant_id, seq DESC),
+  UNIQUE (tenant_id, op_id) -- could also be UNIQUE(op_id)
+);
+```
+
+An operation to the bucket is implemented with a transaction:
+```sql
+-- Generate random op_id.
+BEGIN;
+-- Get the latest state (with the largest sequence number).
+SELECT FOR UPDATE * FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1;
+-- Calculate new state.
+-- Set the new state, using the next sequence number.
+INSERT INTO tenant_usage (tenant_id, seq, ...) (tenant_id, seq+1, ...);
```

Comment:
Worth pointing out that this is a queue-like workload, but the queries won't run into the problems we commonly associate with them.
Also, we're not creating unbounded numbers of MVCC versions for a single key (in fact, never overwrite).
There is a hotspot, so in multi-region settings, depending on where the pods and leaseholders are and how this table is set up, latencies may not be great and there's a throughput ceiling. We should be explicit about what that is and what constraints, if any, it imposes on the frequencies of upgrades.
In particular, it's worth thinking through what will happen if the system cannot sustain the txns in a timely manner at all. You've touched on this in this document ("fall back to assuming 1/N distribution etc") but maybe there's a bit more to the discussion about when to actually start doing that.